lib/revm_consistency_checker/src/revm_state_db.rs

```diff
@@ -0,0 +1,156 @@
+use std::{error::Error, fmt};
+
+use crate::helpers::get_unpadded_code;
+use alloy::primitives::{Address, B256, KECCAK256_EMPTY};
+use reth_revm::{
+    DatabaseRef,
+    db::DBErrorMarker,
+    primitives::{StorageKey, StorageValue},
+    state::{AccountInfo, Bytecode},
+};
+use ruint::aliases::B160;
+use zk_ee::common_structs::derive_flat_storage_key;
+use zk_os_forward_system::run::ReadStorage;
+use zksync_os_interface::{traits::PreimageSource, types::BlockHashes};
+use zksync_os_merkle_tree::fixed_bytes_to_bytes32;
+use zksync_os_storage_api::ReadStateHistory;
+use zksync_os_storage_api::{StateError, ViewState};
+
+#[derive(Debug, Clone)]
+pub struct RevmStateDb<State>
+where
+    State: ReadStateHistory + Clone + Send + 'static,
+{
+    state: State,
+    latest_block: u64,
+    block_hashes: BlockHashes,
+}
+
+impl<State> RevmStateDb<State>
+where
+    State: ReadStateHistory + Clone + Send + 'static,
+{
+    pub fn new(state: State, latest_block: u64, block_hashes: BlockHashes) -> Self {
+        RevmStateDb {
+            state,
+            latest_block,
+            block_hashes,
+        }
+    }
+
+    pub fn set_latest_block(&mut self, latest_block: u64, block_hashes: BlockHashes) {
+        self.latest_block = latest_block;
+        self.block_hashes = block_hashes;
+    }
+}
+
+#[derive(Debug)]
+pub struct RevmStateDbError(anyhow::Error);
+
+// add this:
+impl DBErrorMarker for RevmStateDbError {}
+
+impl fmt::Display for RevmStateDbError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+impl Error for RevmStateDbError {
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        self.0.source()
+    }
+}
+
+// convenient conversions
+impl From<anyhow::Error> for RevmStateDbError {
+    fn from(e: anyhow::Error) -> Self {
+        RevmStateDbError(e)
+    }
+}
+
+impl From<StateError> for RevmStateDbError {
+    fn from(e: StateError) -> Self {
+        RevmStateDbError(e.into())
+    }
+}
+
+impl<State> DatabaseRef for RevmStateDb<State>
+where
+    State: ReadStateHistory + Clone + Send + 'static,
+{
+    /// The database error type.
+    type Error = RevmStateDbError;
```

Comment:
what's the advantage over just    ` type Error = anyhow::Error;`? Alternatively, if you want/need to keep a custom error, you can do 
```
#[derive(Debug, Error)]
#[error(transparent)]
pub struct RevmStateDbError(#[from] anyhow::Error);
``` 
and I think you'll get `Display`, `Error::source` and `From<anyhow::Error>`